src/main/scala/strawman/collection/mutable/ArrayDeque.scala

```diff
@@ -80,75 +93,78 @@ class ArrayDeque[A] private[ArrayDeque](
     this
   }
 
-  override def ++=:(elems: TraversableOnce[A]): this.type = {
+  override def ++=:(elems: TraversableOnce[A]) = {
     // Note this is faster than super.++=: because this does not need to convert TraversableOnce to a Traversable
-    if (elems.isEmpty) return this
-    elems.sizeHintIfCheap match {
-      case srcLength if srcLength >= 0 && 2*(srcLength + this.length) >= mask =>
-        val finalLength = srcLength + this.length
-        val array2 = ArrayDeque.alloc(finalLength)
-        elems.copyToArray(array2.asInstanceOf[Array[A]])
-        copySliceToArray(srcStart = 0, dest = array2, destStart = srcLength, maxItems = size)
-        set(array = array2, start = 0, end = finalLength)
-      case _ =>
-        val copy = clone()
-        end = start
-        this ++= elems ++= copy
+    if (elems.nonEmpty) {
+      ArrayDeque.knownSize(elems) match {
+        case srcLength if srcLength >= 0 && 2*(srcLength + this.length) >= mask =>
+          val finalLength = srcLength + this.length
+          val array2 = ArrayDeque.alloc(finalLength)
+          elems.copyToArray(array2.asInstanceOf[Array[A]])
+          copySliceToArray(srcStart = 0, dest = array2, destStart = srcLength, maxItems = size)
+          resetInternal(array = array2, start = 0, end = finalLength)
+        case _ =>
+          val copy = clone()
+          end = start
+          this ++= elems ++= copy
```

Comment:
FWIW the same idea is used ​in the Spandex PR #52 to get constant time `reverse` (albiet expressed via flipping the `start` and `stop` fields to avoid adding a specific `reversed` field). Only a few operations needed to be updated to take the direction into account, most of the directional logic could be hidden in internal (inlined) methods. It did not affect the benchmarks negatively in any significant way (it did affect the `reverse` benchmark quite positively though 🤓).